File: src/stores/libraryTypes.ts
export type Library = {
  version: Version
  playTime: PlayTime[]
  tracks: TracksHashMap
  trackLists: TrackListsHashMap
}

export type TrackID = string
export type TrackListID = string
export type MsSinceUnixEpoch = number
/** Should be 0-100 */
export type PercentInteger = number
export interface TrackListsHashMap {
  [TrackListID: string]: TrackList
}

export enum Version {
  V1 = 1
}

/** (track id, start time, duration) */
export type PlayTime = [TrackID, number, number]

export interface TracksHashMap {
  [TrackID: string]: Track
}

export type Track = {
  size?: number
  duration?: number
  bitrate?: number
  sampleRate?: number
  file: string
  dateModified: MsSinceUnixEpoch
  dateAdded: MsSinceUnixEpoch
  name?: string
  importedFrom?: string
  /** Imported ID, like iTunes Persistent ID */
  originalId?: string
  artist?: string
  composer?: string
  sortName?: string
  sortArtist?: string
  sortComposer?: string
  genre?: string
  rating?: PercentInteger
  year?: number
  bpm?: number
  comments?: string
  grouping?: string
  liked?: boolean
  disliked?: boolean
  disabled?: boolean
  compilation?: boolean
  albumName?: string
  albumArtist?: string
  sortAlbumName?: string
  sortAlbumArtist?: string
  trackNum?: number
  trackCount?: number
  discNum?: number
  discCount?: number
  dateImported?: MsSinceUnixEpoch
  playCount?: number
  plays?: MsSinceUnixEpoch[]
  playsImported?: CountObject[]
  skipCount?: number
  skips?: MsSinceUnixEpoch[]
  skipsImported?: CountObject[]
  volume?: PercentInteger
}

export type CountObject = {
  count: number
  fromDate: MsSinceUnixEpoch
  toDate: MsSinceUnixEpoch
}

type TrackList = Playlist | Folder | Special

type Playlist = {
  type: "playlist"
  id: TrackListID
  name: string
  description?: string
  liked?: string
  disliked?: string
  importedFrom?: string
  originalId?: string
  dateImported?: MsSinceUnixEpoch
  dateCreated?: MsSinceUnixEpoch
  tracks: TrackID[]
}

type Folder = {
  type: "folder"
  id: TrackListID
  show: boolean
  name: string
  description?: string
  liked?: string
  disliked?: string
  /** For example "itunes" */
  importedFrom?: string
  /** For example iTunes Persistent ID */
  originalId?: string
  dateImported?: MsSinceUnixEpoch
  dateCreated?: MsSinceUnixEpoch
  children: TrackListID[]
}

type Special = {
  type: "special"
  id: TrackListID
  name: SpecialTrackListName
  dateCreated: MsSinceUnixEpoch
  children: TrackListID[]
}

enum SpecialTrackListName {
  Root
}

File: native/src/library_types.rs
#![allow(non_snake_case)]

use linked_hash_map::LinkedHashMap;
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct Library {
  pub version: Version,
  pub tracks: LinkedHashMap<TrackID, Track>,
  pub trackLists: TrackLists,
  pub playTime: Vec<PlayTime>,
}

pub type TrackID = String;
pub type TrackListID = String;
pub type MsSinceUnixEpoch = i64;
/// Should be 0-100
pub type PercentInteger = u8;
pub type TrackLists = LinkedHashMap<TrackListID, TrackList>;

#[derive(Serialize_repr, Deserialize_repr, Debug)]
#[repr(u8)]
pub enum Version {
  V1 = 1,
}

/// (track id, start time, duration)
pub type PlayTime = (TrackID, MsSinceUnixEpoch, i64);

#[derive(Serialize, Deserialize, Debug)]
pub struct Track {
  pub size: i64,
  pub duration: f64,
  pub bitrate: f64,
  pub sampleRate: f64,
  pub file: String,
  pub dateModified: MsSinceUnixEpoch,
  pub dateAdded: MsSinceUnixEpoch,
  pub name: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub importedFrom: Option<String>,
  /// Imported ID, like iTunes Persistent ID:
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub originalId: Option<String>,
  #[serde(default)]
  pub artist: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub composer: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub sortName: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub sortArtist: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub sortComposer: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub genre: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub rating: Option<PercentInteger>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub year: Option<i64>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub bpm: Option<f64>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub comments: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub grouping: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub liked: Option<bool>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub disliked: Option<bool>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub disabled: Option<bool>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub compilation: Option<bool>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub albumName: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub albumArtist: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub sortAlbumName: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub sortAlbumArtist: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub trackNum: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub trackCount: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub discNum: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub discCount: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub dateImported: Option<MsSinceUnixEpoch>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub playCount: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub plays: Option<Vec<MsSinceUnixEpoch>>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub playsImported: Option<Vec<CountObject>>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub skipCount: Option<u32>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub skips: Option<Vec<MsSinceUnixEpoch>>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub skipsImported: Option<Vec<CountObject>>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub volume: Option<PercentInteger>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct CountObject {
  count: i64,
  fromDate: MsSinceUnixEpoch,
  toDate: MsSinceUnixEpoch,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(tag = "type")]
pub enum TrackList {
  #[serde(rename = "playlist")]
  Playlist(Playlist),
  #[serde(rename = "folder")]
  Folder(Folder),
  #[serde(rename = "special")]
  Special(Special),
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Playlist {
  pub id: TrackListID,
  pub name: String,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub description: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub liked: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub disliked: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub importedFrom: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub originalId: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub dateImported: Option<MsSinceUnixEpoch>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub dateCreated: Option<MsSinceUnixEpoch>,
  pub tracks: Vec<TrackID>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Folder {
  pub id: TrackListID,
  pub name: String,
  #[serde(default)]
  pub show: bool,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub description: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub liked: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub disliked: Option<String>,
  /// For example "itunes"
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub importedFrom: Option<String>,
  /// For example iTunes Persistent ID
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub originalId: Option<String>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub dateImported: Option<MsSinceUnixEpoch>,
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub dateCreated: Option<MsSinceUnixEpoch>,
  pub children: Vec<TrackListID>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Special {
  pub id: TrackListID,
  pub name: SpecialTrackListName,
  pub dateCreated: MsSinceUnixEpoch,
  pub children: Vec<TrackListID>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum SpecialTrackListName {
  Root,
}
